### PR TITLE
RFD 0119 - Integrate with AWS using OIDC

### PR DESCRIPTION
**Pull request by marcoandredinis**
[Rendered](https://github.com/gravitational/teleport/blob/marco/rfd_discover_oidc_aws/rfd/0119-aws-api-integration-using-oidc.md)

Part of https://github.com/gravitational/teleport/issues/22129

### CODE REVIEW

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+Turning Teleport into an OIDC IdP, and asking the user to create a new OIDC IdP that consumes the Teleport OIDC endpoints, allows us to easily call AWS API endpoints to, for instance, list RDS instances and their details.
+
+When configuring the provider, we'll need an AWS Role which Teleport uses to issue API Calls.
+
+To store the configuration above, we'll create a new resource Kind: `Integration`.
+It'll leverage the `subkind` prop to distinguish future integrations.
+
+```yaml
+kind: integration
+subkind: aws-oidc
```
Comment (r0mant) [inline rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:59]: Nit: Do we need `-oidc` suffix?

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+version: v1
+metadata:
+	name: some-name
+spec:
+	aws_role: <aws role>
```
Comment (r0mant) [inline rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:64]: Nit:
```suggestion
	aws_role_arn: <aws role>
```

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+## How
+
+AWS allows the [set up](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html) of an IdP using OIDC providers.
+Each provider has a set of roles, which limit their access.
+
+Turning Teleport into an OIDC IdP, and asking the user to create a new OIDC IdP that consumes the Teleport OIDC endpoints, allows us to easily call AWS API endpoints to, for instance, list RDS instances and their details.
+
+When configuring the provider, we'll need an AWS Role which Teleport uses to issue API Calls.
+
+To store the configuration above, we'll create a new resource Kind: `Integration`.
```
Comment (r0mant) [inline rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:54]: Can users manage these integrations dynamically with tctl or API? I assume they can, let's include some examples.

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+AWS receives the request and validates the token against the public key provided by OIDC IdP (ie, Teleport) at the `jwks_uri`.
+If authenticated and authorized for the api call, AWS returns the response to the API call.
+
+### Implementation Details
+
+#### Signing Key
+One of the requirements to be an OIDC provider is to provide the public key in a known HTTP endpoint and sign a JSON object (with claims) with the private key.
+
+We'll create a new RSA 2048 Key type and add it to the CertAuthority.
+This will be very similar to the JWTSigner key type used for App access.
```
Comment (r0mant) [inline rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:139]: So are we going to introduce another CA type? Like, `OIDCCA`? Mention it explicitly in the document.

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+  "jwks_uri": "<teleportProviderURL>/.well-known/jwks-oidc",
+  "claims": ["iss", "sub", "aud", "jti", "iat", "exp", "nbf"],
+  "id_token_signing_alg_values_supported": ["RS256"],
+  "response_types_supported": ["id_token"],
+  "scopes_supported": ["openid"],
+  "subject_types_supported": ["public", "pairwise"]
+}
+```
+
+- Issuer: the provider's URL
```
Comment (r0mant) [inline rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:164]: This is basically proxy's public address, correct? Let's just call it that explicitly so it's clear.

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+- JWKS URI: the endpoint where the provider returns the public keys.
+- Claims Supported: a list of supported claims that the OpenID Provider (ie, Teleport) MAY be able to supply values for when issuing a token:
+  - `iss`: the issuer identity, on our case it will be the same as the `issuer` key.
+  - `sub`: identifies the subject of the token, on our case it will contain the user's name.
```
Comment (r0mant) [inline rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:168]:
```suggestion
  - `sub`: identifies the subject of the token, in our case it will contain the Teleport username.
```

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+  - `aud`: the audience for the token, on our case it will be the same as the `audience` defined when configuring the OIDC IdP in AWS (step 3.3 for the User's Point of View).
+  - `jti`: is the token id.
```
Comment (r0mant) [inline rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:170]: UUIDv4? Mention that.

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+The AWS Role must trust Teleport as an IdP.
+To do so, the user must add the following Trusted relationship:
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Principal": {
+                "Federated": "arn:aws:iam::<account id>:oidc-provider/<teleportProviderURL>"
```
Comment (r0mant) [inline rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:246]: Do you embed the actual proxy address here and below? Can you use specific values for better demonstration? E.g. say we have a proxy at `https://proxy.example.com:3080`.

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+
+#### Rotation
+JWT Signing keys can be rotated using `tctl auth rotate --type=oidc`.
+
+JWKS endpoint returns a list of public keys, so the old and the new key are provided for a seamless migration.
+
+#### MITM between AWS and Teleport
+
+The only configuration AWS has about the provider is its URL.
+If that DNS is controlled by some evil part, AWS might accept requests from that evil part.
```
Comment (r0mant) [inline rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:269]:
```suggestion
If DNS is controlled by an attacker, AWS might accept requests from them.
```

**Review by r0mant — COMMENTED** [review]
First pass. @mdwn Could you take a look as well when you get a chance, in particular at OIDC-related bits?

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md [hunk 10]
```diff
@@ -0,0 +1,495 @@
...
+* Product: @xinding33 || @klizhentas
+* Security: @reedloden
+
+## What
+
+Discover AWS RDS Databases without manual configuration.
+
+#### Goals
+
+Easily discover AWS RDS Databases of an account.
```
[the remark this answers is not in the document]
Comment (marcoandredinis) [reply rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:20]: I was targeting something more specific to avoid the "where are we going to use this"

Let me rephrase the goal and RFD title to "Integrate with AWS using OIDC"

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md (hunk 1 above)
Comment (marcoandredinis) [reply rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:59]: I'm trying to leave space/pattern to have other types of AWS integrations (using secrets or using an ec2 instance).

Should I remove the `-oidc` anyway?

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md (hunk 4 above)
Comment (strideynet) [reply rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:139]: I'd be keen for us to introduce a CA just for issuing OIDC ID Tokens

**Review by strideynet — COMMENTED** [review]
I'm really excited for an OIDC IdP starting to be added to Teleport. This will eventually unlock the ability for Teleport to issue ID tokens to processes running on hosts with Machine ID installed in order for them to be able to access cloud APIs.

One thing I would consider would be actually issuing the token in the name of the Teleport proxy rather than the user for the purposes of AWS RDS DB discovery - especially if the Proxy itself is going to actually be making the call. This accounts for a scenario in future where Teleport is also issuing tokens directly to users, but they do not want to grant the ability to all their users the ability to list RDS DBs. For example: 

`sub`: `system:proxy:discover` 

This allows the configured trusted relationship to be explicitly for the discover flow.

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md [hunk 11]
```diff
@@ -0,0 +1,495 @@
...
+* IdP: Identity Provider is a system that manages identities.
+* JWT: JSON Web Token is a json encoded token that contains a set of claims signed by an entity.
+* JWKS: JSON Web Key Set is a set of public keys that IdP consumers must use to validate a signed JWT.
+
+## Why
+
+Our current discovery process for AWS resources requires multiple steps before the user gets to the "aha moment".
+Some of those steps are a little complicated and require the user to know some Teleport specific internals.
+
+We can reduce the number of steps and remove most of the Teleport specific configuration by creating an OIDC IdP in Teleport.
```
[the remark this answers is not in the document]
Comment (marcoandredinis) [reply rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:43]: :+1:
Mentioning cloud here seems out of place :thinking: 

I'll add it to the goals instead, if that's ok

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md (hunk 3 above)
Comment (marcoandredinis) [reply rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:54]: I was thinking of only supporting via API, but I guess we should add `tctl` as well (it's always handy and not very expensive)

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md [hunk 12]
```diff
@@ -0,0 +1,495 @@
...
+If that DNS is controlled by some evil part, AWS might accept requests from that evil part.
+
+However, assuming the DNS is controlled by an evil part, then the users that are logging in to Teleport are also providing their credentials to that part.
+
+We don't think this is a scenario we should focus on this RFD.
+
+#### AWS Calls from unauthorized Teleport Users
+We'll be able to call `DescribeDBInstances` AWS endpoint, however we must ensure that this is protected by RBAC.
+
+In order to do so, we'll re-use the `db` resource with the `list` verb to allow listing RDS Databases.
```
[the remark this answers is not in the document]
Comment (marcoandredinis) [reply rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:278]: Picture the state where we have an Integration with AWS.
We'll have an HTTP endpoint in Proxy that calls the AWS API Describe DB Instance.
We'll have a new resource Integration, and subsequent rules (create, delete, list, update), but that only works for CRUD operations over the resource itself.

Should any user be able to list the AWS RDS DBs? Or do we want some kind of check before calling the AWS Endpoint?
I'm overloading the `db.list` rule with something else not entirely relevant.
However, I think we should add some check before calling the AWS Endpoint.

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcoandredinis) [reply rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md:168]: Sure, seems fair :+1:

**Comment by marcoandredinis** [comment]
>One thing I would consider would be actually issuing the token in the name of the Teleport proxy rather than the user for the purposes of AWS RDS DB discovery - especially if the Proxy itself is going to actually be making the call. This accounts for a scenario in future where Teleport is also issuing tokens directly to users, but they do not want to grant the ability to all their users the ability to list RDS DBs. For example:

Proxy is making the call but the user was the one who triggered it :thinking: 
We can set the `sub` to be the system (proxy/discover) but I think we should also somehow identify the user.
We can always add another claim :thinking:

**Comment by strideynet** [comment]
> > One thing I would consider would be actually issuing the token in the name of the Teleport proxy rather than the user for the purposes of AWS RDS DB discovery - especially if the Proxy itself is going to actually be making the call. This accounts for a scenario in future where Teleport is also issuing tokens directly to users, but they do not want to grant the ability to all their users the ability to list RDS DBs. For example:
> 
> Proxy is making the call but the user was the one who triggered it 🤔 We can set the `sub` to be the system (proxy/discover) but I think we should also somehow identify the user. We can always add another claim 🤔

Perhaps a "on-behalf-of" claim or similar ? I'm unsure, but I do think that identifying the proxy discover subsystem lets us craft a much more specific trust rule in AWS.

File: rfd/9999-aws-api-integration-using-oidc.md [hunk 13]
```diff
@@ -0,0 +1,554 @@
...
+                        │  └────►│ Service   ├──┼─────────────────────┘
+                        │        │ [ RDS ]   │  │
+                        │        └───────────┘  │
+                        │                       │
+                        └───────────────────────┘
+```
+
+#### User's Point of View
+When the user adds a new resource, using the Discover Wizard, they'll be asked where the resource lives.
+Choosing an RDS DB, the user is presented with the following flow:
```
Comment (r0mant) [inline rfd/9999-aws-api-integration-using-oidc.md:101]: The "Add AWS integration" flow can be gone through separately from the Discover wizard though too, right? And it will be skipped if you already have an integration setup? So I'd probably rephrase this to something like "When a users adds a new AWS integration, they will go through the following. Users will also be asked to set up an AWS integration when adding RDS databases in Discover flow."

File: rfd/9999-aws-api-integration-using-oidc.md [hunk 14]
```diff
@@ -0,0 +1,554 @@
...
+
+At this point, the integration is configured and the user is able to discover RDS DBs without leaving Teleport.
+
+#### System's Point of View
+The first interaction between AWS and Teleport happens when the user clicks on `Get Thumbprint` (step 4 of the User's Point of View flow).
+This will trigger a request started by AWS to Teleport hitting the OpenID Configuration endpoint and subsequently the JSON Web Key Set endpoint.
+
+After this step, when the user tries to list RDS DBs, Teleport generates a token and issues the API call.
+
+To generate a token, Teleport creates a JWT with the claims described above and the configured role, and signs it with the private key (the public is provided in the public JWKS endpoint).
```
Comment (r0mant) [inline rfd/9999-aws-api-integration-using-oidc.md:126]: > JWT with the claims described above

What are these claims? I don't see any claims being described above.

File: rfd/9999-aws-api-integration-using-oidc.md [hunk 15]
```diff
@@ -0,0 +1,554 @@
...
+
+To ease the necessary manual steps, the Discover wizard will provide a small video demoing what the user must do for the previous steps.
+
+At this point, the integration is configured and the user is able to discover RDS DBs without leaving Teleport.
+
+#### System's Point of View
+The first interaction between AWS and Teleport happens when the user clicks on `Get Thumbprint` (step 4 of the User's Point of View flow).
+This will trigger a request started by AWS to Teleport hitting the OpenID Configuration endpoint and subsequently the JSON Web Key Set endpoint.
+
+After this step, when the user tries to list RDS DBs, Teleport generates a token and issues the API call.
```
Comment (r0mant) [inline rfd/9999-aws-api-integration-using-oidc.md:124]: Can you have multiple AWS integrations e.g. for different accounts? How do we pick which one to use?

File: rfd/9999-aws-api-integration-using-oidc.md [hunk 16]
```diff
@@ -0,0 +1,554 @@
...
+version: v1
+metadata:
+	name: some-name
+spec:
+	aws_role_arn: arn:aws:123:TeleportOIDC
+
+$ tctl create aws-integration.yaml
+```
+
+Resource representation as `yaml`:
```
Comment (r0mant) [inline rfd/9999-aws-api-integration-using-oidc.md:307]: Can you add a section covering IaC flows to the RFD as well? Are there any changes needed to Terraform provider, Kube operator and Helm charts?

File: rfd/9999-aws-api-integration-using-oidc.md [hunk 17]
```diff
@@ -0,0 +1,554 @@
...
+metadata:
+  name: editor
+spec:
+  allow:
+    rules:
+    - resources:
+      - db
+      verbs:
+      - list
+```
```
Comment (r0mant) [inline rfd/9999-aws-api-integration-using-oidc.md:350]: I still don't get this tbh. In your comment you said:

> We'll have an HTTP endpoint in Proxy that calls the AWS API Describe DB Instance.

What endpoint is this and why would it be unauthenticated?

File: rfd/9999-aws-api-integration-using-oidc.md [hunk 18]
```diff
@@ -0,0 +1,554 @@
...
+After this step, when the user tries to list RDS DBs, Teleport generates a token and issues the API call.
+
+To generate a token, Teleport creates a JWT with the claims described above and the configured role, and signs it with the private key (the public is provided in the public JWKS endpoint).
+
+AWS receives the request and validates the token against the public key provided by OIDC IdP (ie, Teleport) at the `jwks_uri`.
+If authenticated and authorized for the api call, AWS returns the response to the API call.
+
+### Implementation Details
+
+#### Signing Key
```
Comment (strideynet) [inline rfd/9999-aws-api-integration-using-oidc.md:133]: Can we call out what type of signing key we will use here? RS256 ?

File: rfd/9999-aws-api-integration-using-oidc.md [hunk 19]
```diff
@@ -0,0 +1,554 @@
...
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Principal": {
+                "Federated": "arn:aws:iam::0123456789012:oidc-provider/proxy.example.com"
+            },
+            "Action": "sts:AssumeRoleWithWebIdentity",
+            "Condition": {
+                "StringEquals": {
+                    "proxy.example.com:aud": "sts.amazonaws.com"
```
Comment (strideynet) [inline rfd/9999-aws-api-integration-using-oidc.md:251]: If we used an ID token subject that was specific to the Teleport Discovery process then we could specify this here to ensure that this role can only be used by that process and not by other entities which have been issued tokens.

File: rfd/9999-aws-api-integration-using-oidc.md [hunk 20]
```diff
@@ -0,0 +1,554 @@
...
+
+AWS receives the request and validates the token against the public key provided by OIDC IdP (ie, Teleport) at the `jwks_uri`.
+If authenticated and authorized for the api call, AWS returns the response to the API call.
+
+### Implementation Details
+
+#### Signing Key
+One of the requirements to be an OIDC provider is to provide the public key in a known HTTP endpoint and sign a JSON object (with claims) with the private key.
+
+We'll create a new CA: OIDCCA.
```
Comment (mdwn) [inline rfd/9999-aws-api-integration-using-oidc.md:136]: There's a reference SAML IdP CA that's new as well. May be a better reference point as it does exactly the same thing.

File: rfd/9999-aws-api-integration-using-oidc.md [hunk 21]
```diff
@@ -0,0 +1,554 @@
...
+#### User's Point of View
+When the user adds a new resource, using the Discover Wizard, they'll be asked where the resource lives.
+Choosing an RDS DB, the user is presented with the following flow:
+
+1. User is prompted to setup an OIDC Identity Provider pointing to its Teleport instance's url.
+2. User opens AWS, selects IAM and then, under `Access Management`, selects `Identity providers`.
+3. Configures the Teleport instance as an OIDC IdP:
+   1. Clicks `Add provider` of type `OpenID Connect`.
+   2. Fills in the Provider URL as provided by the Discover Wizard.
+   3. Fills in the Audience as provided by the Discover Wizard `sts.amazonaws.com`.
```
Comment (mdwn) [inline rfd/9999-aws-api-integration-using-oidc.md:108]: Does the audience have to be `sts.amazonaws.com`? We could skip this step if we had a fixed audience as part of the OIDC IdP in Teleport.

**Review by mdwn — COMMENTED** [review]
We should specifically call out which auth flows we support for this OIDC. One thing I ran into while experimenting with this is back channel vs. front channel auth. Back channel meaning that there has to be direct communication between the IdP and the receiving service, which may or may not be the case depending on a user setup.

I mention this because Okta in particular is opinionated about the use of the `id_token` flow, which could be a problem when generalizing this.

Handy reference; https://www.scottbrady91.com/openid-connect/openid-connect-flows

File: rfd/9999-aws-api-integration-using-oidc.md (hunk 21 above)
Comment (strideynet) [reply rfd/9999-aws-api-integration-using-oidc.md:108]: Whilst AWS doesn't enforce it, and I'd argue they should, `sts.amazonaws.com` definitely makes the most semantic sense for the value here as that's the host of the AWS service that allows the OIDC ID token to be swapped for AWS credentials. I imagine they don't enforce it universally as not too long ago some CI/CD providers did not allow you to customise the audience of issued tokens to workloads.

File: rfd/9999-aws-api-integration-using-oidc.md (hunk 14 above)
Comment (marcoandredinis) [reply rfd/9999-aws-api-integration-using-oidc.md:126]: They are below and not above, sorry.
Replaced with `below` with a markdown link to it.

File: rfd/9999-aws-api-integration-using-oidc.md (hunk 15 above)
Comment (marcoandredinis) [reply rfd/9999-aws-api-integration-using-oidc.md:124]: Yes, multiple AWS Integrations must be supported.
I'll add that.

I'll also add that the user must be able to select which AWS Integration to use.

File: rfd/9999-aws-api-integration-using-oidc.md (hunk 16 above)
Comment (marcoandredinis) [reply rfd/9999-aws-api-integration-using-oidc.md:307]: Yes :+1:

File: rfd/9999-aws-api-integration-using-oidc.md (hunk 17 above)
Comment (marcoandredinis) [reply rfd/9999-aws-api-integration-using-oidc.md:350]: There's still authentication, however my concern is authorization.

Let's say we have a UserA that has `editor` role and is able to List AWS RDS DBs.

However, there's also UserB that has `access` role.
Should UserB be allowed to call the List AWS RDS DBs?

File: rfd/9999-aws-api-integration-using-oidc.md (hunk 18 above)
Comment (marcoandredinis) [reply rfd/9999-aws-api-integration-using-oidc.md:133]: Yes :+1:

File: rfd/9999-aws-api-integration-using-oidc.md (hunk 19 above)
Comment (marcoandredinis) [reply rfd/9999-aws-api-integration-using-oidc.md:251]: Something like this?
```json
            "Condition": {
                "StringEquals": {
                    "proxy.example.com:aud": "sts.amazonaws.com",
                    "proxy.example.com:sub": "system:proxy:discover",
```

Assuming we issue tokens whose `sub` is `system:proxy:discover`.

What would be the pros of this approach?
Allow for multiple roles to be associated with the same provider but having different policies.

If we are strict here, then we would need users to create extra roles in order to have multiple Teleport "systems" interacting with it.

Honestly, I think we should not add that constraint to not increase scope.
We can always add that condition later on if needed.

Comment (ryanclark) [reply rfd/9999-aws-api-integration-using-oidc.md:251]: I'd rather avoid this as it requires further setup to the trust relationship (which is created automatically in the IdP setup flow) with not much added benefit

File: rfd/9999-aws-api-integration-using-oidc.md (hunk 17 above)
Comment (marcoandredinis) [reply rfd/9999-aws-api-integration-using-oidc.md:414]: I added some more details on Listing AWS RDS DBs
https://github.com/gravitational/teleport/pull/22556/commits/83c735ba94cf63b03ad589b2f0e617e0fcda9068

Let me know if that helps in understanding my concern

File: rfd/9999-aws-api-integration-using-oidc.md [hunk 22]
```diff
@@ -0,0 +1,618 @@
...
+The user will create or associate an AWS Role to the Teleport OIDC IdP.
+
+As an example, we the following is the required policy to list RDS DBs (including Aurora clusters):
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Actions": "[
```
Comment (strideynet) [inline rfd/9999-aws-api-integration-using-oidc.md:234]:
```
"Action": [
  "rds:DescribeDBClusters",
  "rds:DescribeDBInstances"
]

File: rfd/9999-aws-api-integration-using-oidc.md (hunk 19 above)
Comment (strideynet) [reply rfd/9999-aws-api-integration-using-oidc.md:258]: Mhm, I can definitely see the concern from the UX point of view. 

But, I really think as part of the RFD we need to acknowledge that we are opening up access to list all RDS DBs to any machine or user in future if an operator misses an update advisory and does not modify their existing trust relationships created before we issued the change in advice. Whilst this is certainly a low privilege action, one concern I have is that the design from this RfD is used with another feature that requires more privileged permissions and because this wasn't openly acknowledged in this RfD, that this will be missed.

Comment (marcoandredinis) [reply rfd/9999-aws-api-integration-using-oidc.md:258]: We can add
```
 "StringEquals": {
                    "proxy.example.com:aud": "sts.amazonaws.com",
                    "proxy.example.com:sub": "system:proxy",
```

And if we ever need to evolve that, we can just add another suffix to the `sub`.
What do you think?

Comment (ryanclark) [reply rfd/9999-aws-api-integration-using-oidc.md:258]: @strideynet I might be misinterpreting your concern, let's chat on Slack if I'm off -

> But, I really think as part of the RFD we need to acknowledge that we are opening up access to list all RDS DBs to any machine or user in future if an operator misses an update advisory and does not modify their existing trust relationships created before we issued the change in advice

Hmm, I don't see this happening - the AWS provider is a one-time setup with an IAM policy, that grants the Teleport auth service access to AWS, which can only be accessed through the AWS integration. This mechanism would be the only thing that sets the `sub` when signing the OIDC JWT, so I don't see the advantage in adding a condition that's always going to evaluate to true

> Whilst this is certainly a low privilege action, one concern I have is that the design from this RfD is used with another feature that requires more privileged permissions and because this wasn't openly acknowledged in this RfD, that this will be missed.

Like above, there's only one feature that's going to use this

Comment (ryanclark) [reply rfd/9999-aws-api-integration-using-oidc.md:258]: @marcoandredinis - @strideynet and I had a chat last night about this section and came to the compromise of setting the audience to something other than `sts.amazonaws.com` - maybe `discover.teleport` or `discover.proxy-url`

As per the [AWS docs](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html#manage-oidc-provider-console) the audience is meant to be the client ID of the application, so setting it to something Discover related is more suitable anyway.

This means we can avoid setting a subject as we're not going to conflict with any future developments of the IdP provider (they'll be requesting a different audience).

File: rfd/9999-aws-api-integration-using-oidc.md [hunk 23]
```diff
@@ -0,0 +1,621 @@
...
+JWKS endpoint returns a list of public keys, so the old and the new key are provided for a seamless migration.
+
+#### MITM between AWS and Teleport
+
+The only configuration AWS has about the provider is its URL.
+If DNS is controlled by an attacker, AWS might accept requests from them.
+
+However, assuming the DNS is controlled by an attacker, then the users that are logging in to Teleport are also providing their credentials to that part.
+
+We don't think this is a scenario we should focus on this RFD.
```
Comment (strideynet) [inline rfd/9999-aws-api-integration-using-oidc.md:396]: This is somewhat mitigated by AWS's thumb printing of the ID provider's certificate's top intermediate certificate authority - an attacker would need to poison AWS's DNS resolvers and also obtain a legitimate certificate issued by the same intermediate CA.

I agree with your overall assessment here. If AWS's DNS resolvers have been poisoned and their chain of trust compromised, that's a much bigger problem that's out of our hands no matter what we do.

**We should really add a warning that if you change your certificate provider that issues your Teleport proxy certificate, a new fingerprint must be added to AWS.**

File: rfd/9999-aws-api-integration-using-oidc.md [hunk 24]
```diff
@@ -0,0 +1,621 @@
...
+### Security
+
+#### Rotation
+JWT Signing keys can be rotated using `tctl auth rotate --type=oidc`.
+
+JWKS endpoint returns a list of public keys, so the old and the new key are provided for a seamless migration.
+
+#### MITM between AWS and Teleport
+
+The only configuration AWS has about the provider is its URL.
```
Comment (strideynet) [inline rfd/9999-aws-api-integration-using-oidc.md:391]: As noted in my lower comment, this isn't strictly true, AWS also fingerprints the top intermediate certificate authority that issued the IdP's TLS certificate.

**Review by strideynet — APPROVED** [review]
Looks good - one small comment on the security section that would nice to be resolved before merging.

File: rfd/9999-aws-api-integration-using-oidc.md (hunk 23 above)
Comment (ryanclark) [reply rfd/9999-aws-api-integration-using-oidc.md:396]: Definitely. I noticed this earlier, from the AWS IdP Page

> Server certificate thumbprint is the hex-encoded SHA-1 hash value of the X.509 certificate used by the domain where the OpenID Connect provider makes its keys available.

We should aim to mitigate this somehow, be it via warning or an alert

Comment (marcoandredinis) [reply rfd/9999-aws-api-integration-using-oidc.md]: I added some notes here
https://github.com/gravitational/teleport/pull/22556/commits/583d8b3f8d18ad7678f6c1b7848505c23cc402c9

> be it via warning or an alert

Teleport alerting the user?
:thought_balloon: 
We can store our own thumbprint within the integration.
If we get an error from AWS (403?), we validate if the thumbprint is still the same.
If they are not the same, then warn the user about it (and that they need to set up the integration again).

Comment (strideynet) [reply rfd/9999-aws-api-integration-using-oidc.md]: I think a warning in the docs somewhere may suffice for now, but, if we were able to prompt in the logs on failure that would be awesome :')

**Review by strideynet — APPROVED** [review]

File: rfd/9999-aws-api-integration-using-oidc.md (hunk 23 above)
Comment (ryanclark) [reply rfd/9999-aws-api-integration-using-oidc.md]: I was thinking maybe a cluster alert if the fingerprint has changed

File: rfd/9999-aws-api-integration-using-oidc.md [hunk 25]
```diff
@@ -0,0 +1,640 @@
...
+
+`tctl`:
+
+```
+$ tctl get integrations
+kind: integration
+subkind: aws-oidc
+version: v1
+metadata:
+	name: some-name
```
Comment (r0mant) [inline rfd/9999-aws-api-integration-using-oidc.md:358]: Here and everywhere below in yaml resources indentation seems off. Use spaces instead of tabs?

**Review by r0mant — APPROVED** [review]

File: rfd/9999-aws-api-integration-using-oidc.md (hunk 25 above)
Comment (marcoandredinis) [reply rfd/9999-aws-api-integration-using-oidc.md]: :thinking: I think it was correct but using 4 spaces instead of the usual (?) 2
I also added the proto representation of the Integration resource

**Comment by r0mant** [comment]
@marcoandredinis Btw, don't forget to update the RFD number in the PR title and RFD itself please.

**Review by klizhentas — APPROVED** [review]

**Comment by klizhentas** [comment]
@r0mant I'd like to emphasize that I think this flow will work great for smaller teams that just start with their AWS account. 

We must have a parallel IAC terraform flow for more mature teams that would like to configure everything without magic declaratively.

File: rfd/0119-aws-api-integration-using-oidc.md [hunk 26]
```diff
@@ -0,0 +1,686 @@
...
+
+    // Integration is not running because it was just created or was disabled by the user.
+    INTEGRATION_STATUS_PAUSED = 1;
+
+    // Integration is running.
+    INTEGRATION_STATUS_RUNNING = 2;
+
+    // Integration has an error and that should be fixed before receiving requests.
+    INTEGRATION_STATUS_ERROR = 3;
+  }
```
[the remark this answers is not in the document]
Comment (marcoandredinis) [reply rfd/0119-aws-api-integration-using-oidc.md:96]: The integration may become unhealthy (OIDC IdP was removed from AWS, Role doesn't allow for required api calls, thumbprint changed).
In that case, there will be a Cluster Alert and the Integration will switch to Error Status.

The Paused status is to allow the user to disable the Integration but not removing it.
Maybe that's something we can work later on, and is not really needed for this first iteration.

[the remark this answers is not in the document]
Comment (ryanclark) [reply rfd/0119-aws-api-integration-using-oidc.md:96]: If the proxy rotates its SSL certificate the integration will also become unhealthy as AWS saves and compares the fingerprint of the SSL certificate on setup

[the remark this answers is not in the document]
Comment (ryanclark) [reply rfd/0119-aws-api-integration-using-oidc.md:96]: thumbprint*

[the remark this answers is not in the document]
Comment (strideynet) [reply rfd/0119-aws-api-integration-using-oidc.md:96]: It's not quite that drastic, it actually uses the fingerprint of the TLS certificate's CA. So you can rotate certificates as long as it's issued by the same CA.

[the remark this answers is not in the document]
Comment (ryanclark) [reply rfd/0119-aws-api-integration-using-oidc.md:96]: Ah I misunderstood. Cheers

[the remark this answers is not in the document]
Comment (r0mant) [reply rfd/0119-aws-api-integration-using-oidc.md:96]: @marcoandredinis I understand the general intent but tbh it just raises more questions: how would Teleport detect that it's unhealthy and why, and whether it's a permanent or temporary unhealthiness, and would we detect it's healthy after being unhealthy, and what does the user need to do in that case? And so on.

We did have some ideas (or rather, "aspirations") for upstream resource health checks (e.g. https://github.com/gravitational/teleport/issues/20544) but it feels like it's a large enough problem that would require a separate RFD and consistent approach we can use for all resources.

I would cut this from the scope of this RFD and implementation for now.

[the remark this answers is not in the document]
Comment (marcoandredinis) [reply rfd/0119-aws-api-integration-using-oidc.md:96]: Sounds fair :+1: 
I'll remove the status from the RFD and implementation (backport to V12 is not merged yet, so it will wait to include this removal)

File: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (ravicious) [reply rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md]: > We'll have a new verb: `use` If the user has `integration.use` then they can issue Integration API Calls

Apologies for a naive, after-the-merge question, but what is stopping us from using one of the existing verbs like "read" for this? Does it have to do with the part where you mention rules working only for CRUD operations over the resource itself?

We have managed to avoid adding a verb like "use" for resources where one could theoretically make use of it as well ("using" a db server or a node) so I'm wondering how this situation is different.

If it's related to the `POST .../integration/:id/action/aws-oidc-list-databases` endpoint mentioned in the RFD, then we could even argue that describing access to call an integration action could be made through checking the "create" verb for a new resource called "integration action". OTOH I've seen someone mention that it's hard to implement those checks for "virtual" resources that don't actually exist on the backend.

[the remark this answers is not in the document]
Comment (marcoandredinis) [reply rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md]: There's some discussion here, which provide some context
https://github.com/gravitational/teleport/pull/24256#discussion_r1163043822

> We have managed to avoid adding a verb like "use" for resources where one could theoretically make use of it as well ("using" a db server or a node) so I'm wondering how this situation is different.

They are protected by `<resource>_labels`, so we could also use that here.
We discussed that approach but ended deciding on using the `VerbUse`

As for creating a virtual resource `integration action`, I don't think it was considered.
I'm not aware of any issues of implementing resources without a backend. But I didn't try it.

I think it could be an alternative, but don't actually see a big disadvantage in using the `VerbUse`
